Subject: [PATCH] Port QPR2 window insets animation observers

Port the public listener, registration/removal, observer merge and
exception-safe dispatch from the pinned QPR2 GSI. Wire prepare and end
at the actual InsetsController runner lifecycle, including cancellation.
Enable this observer lifecycle in the explicit QPR2 profile. Keep all
Samsung methods and existing animation callbacks; no native replacement.

diff --git a/smali_classes4/android/view/InsetsController$Host.smali b/smali_classes4/android/view/InsetsController$Host.smali
--- a/smali_classes4/android/view/InsetsController$Host.smali
+++ b/smali_classes4/android/view/InsetsController$Host.smali
@@ -157,3 +157,15 @@
 
     return p0
 .end method
+
+.method public blacklist dispatchWindowInsetsAnimationEndToObserver(Landroid/view/WindowInsetsAnimation;)V
+    .locals 0
+
+    return-void
+.end method
+
+.method public blacklist dispatchWindowInsetsAnimationPrepareToObserver(Landroid/view/WindowInsetsAnimation;)V
+    .locals 0
+
+    return-void
+.end method
diff --git a/smali_classes4/android/view/InsetsController.smali b/smali_classes4/android/view/InsetsController.smali
--- a/smali_classes4/android/view/InsetsController.smali
+++ b/smali_classes4/android/view/InsetsController.smali
@@ -1025,8 +1025,20 @@
     :cond_5
     move v0, v3
 
+    goto :unica_insets_cleanup
+
     :cond_6
     :goto_4
+    # Only a runner removed from mRunningAnimations receives this end event.
+    iget-object p2, p0, Landroid/view/InsetsController;->mHost:Landroid/view/InsetsController$Host;
+
+    invoke-interface {p1}, Landroid/view/InsetsAnimationControlRunner;->getAnimation()Landroid/view/WindowInsetsAnimation;
+
+    move-result-object v4
+
+    invoke-interface {p2, v4}, Landroid/view/InsetsController$Host;->dispatchWindowInsetsAnimationEndToObserver(Landroid/view/WindowInsetsAnimation;)V
+
+    :unica_insets_cleanup
     if-lez v0, :cond_9
 
     iget p2, p0, Landroid/view/InsetsController;->mAnimatingTypes:I
@@ -1860,6 +1872,15 @@
     invoke-direct {v2, v0, v9}, Landroid/view/InsetsController$RunningAnimation;-><init>(Landroid/view/InsetsAnimationControlRunner;I)V
 
     invoke-virtual {v1, v2}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    # QPR2 observer lifecycle is enabled for this explicit Paradigm profile.
+    iget-object v1, v7, Landroid/view/InsetsController;->mHost:Landroid/view/InsetsController$Host;
+
+    invoke-interface {v0}, Landroid/view/InsetsAnimationControlRunner;->getAnimation()Landroid/view/WindowInsetsAnimation;
+
+    move-result-object v2
+
+    invoke-interface {v1, v2}, Landroid/view/InsetsController$Host;->dispatchWindowInsetsAnimationPrepareToObserver(Landroid/view/WindowInsetsAnimation;)V
 
     sget-object v1, Landroid/view/ViewProtoLogGroups;->INSETS_CONTROLLER_DEBUG:Lcom/android/internal/protolog/ProtoLogGroup;
 
diff --git a/smali_classes4/android/view/ViewRootInsetsControllerHost.smali b/smali_classes4/android/view/ViewRootInsetsControllerHost.smali
--- a/smali_classes4/android/view/ViewRootInsetsControllerHost.smali
+++ b/smali_classes4/android/view/ViewRootInsetsControllerHost.smali
@@ -704,3 +704,45 @@
 
     return p0
 .end method
+
+.method public blacklist dispatchWindowInsetsAnimationEndToObserver(Landroid/view/WindowInsetsAnimation;)V
+    .locals 1
+
+    iget-object v0, p0, Landroid/view/ViewRootInsetsControllerHost;->mViewRoot:Landroid/view/ViewRootImpl;
+
+    iget-object v0, v0, Landroid/view/ViewRootImpl;->mView:Landroid/view/View;
+
+    if-eqz v0, :cond_0
+
+    iget-object p0, p0, Landroid/view/ViewRootInsetsControllerHost;->mViewRoot:Landroid/view/ViewRootImpl;
+
+    iget-object p0, p0, Landroid/view/ViewRootImpl;->mAttachInfo:Landroid/view/View$AttachInfo;
+
+    iget-object p0, p0, Landroid/view/View$AttachInfo;->mTreeObserver:Landroid/view/ViewTreeObserver;
+
+    invoke-virtual {p0, p1}, Landroid/view/ViewTreeObserver;->dispatchOnWindowInsetsAnimationEnd(Landroid/view/WindowInsetsAnimation;)V
+
+    :cond_0
+    return-void
+.end method
+
+.method public blacklist dispatchWindowInsetsAnimationPrepareToObserver(Landroid/view/WindowInsetsAnimation;)V
+    .locals 1
+
+    iget-object v0, p0, Landroid/view/ViewRootInsetsControllerHost;->mViewRoot:Landroid/view/ViewRootImpl;
+
+    iget-object v0, v0, Landroid/view/ViewRootImpl;->mView:Landroid/view/View;
+
+    if-eqz v0, :cond_0
+
+    iget-object p0, p0, Landroid/view/ViewRootInsetsControllerHost;->mViewRoot:Landroid/view/ViewRootImpl;
+
+    iget-object p0, p0, Landroid/view/ViewRootImpl;->mAttachInfo:Landroid/view/View$AttachInfo;
+
+    iget-object p0, p0, Landroid/view/View$AttachInfo;->mTreeObserver:Landroid/view/ViewTreeObserver;
+
+    invoke-virtual {p0, p1}, Landroid/view/ViewTreeObserver;->dispatchOnWindowInsetsAnimationPrepare(Landroid/view/WindowInsetsAnimation;)V
+
+    :cond_0
+    return-void
+.end method
diff --git a/smali_classes4/android/view/ViewTreeObserver$WindowInsetsAnimationListener.smali b/smali_classes4/android/view/ViewTreeObserver$WindowInsetsAnimationListener.smali
new file mode 100644
--- /dev/null
+++ b/smali_classes4/android/view/ViewTreeObserver$WindowInsetsAnimationListener.smali
@@ -0,0 +1,28 @@
+.class public interface abstract Landroid/view/ViewTreeObserver$WindowInsetsAnimationListener;
+.super Ljava/lang/Object;
+.source "ViewTreeObserver.java"
+
+
+# annotations
+.annotation system Ldalvik/annotation/EnclosingClass;
+    value = Landroid/view/ViewTreeObserver;
+.end annotation
+
+.annotation system Ldalvik/annotation/InnerClass;
+    accessFlags = 0x609
+    name = "WindowInsetsAnimationListener"
+.end annotation
+
+
+# virtual methods
+.method public whitelist onEnd(Landroid/view/WindowInsetsAnimation;)V
+    .locals 0
+
+    return-void
+.end method
+
+.method public whitelist onPrepare(Landroid/view/WindowInsetsAnimation;)V
+    .locals 0
+
+    return-void
+.end method
diff --git a/smali_classes4/android/view/ViewTreeObserver.smali b/smali_classes4/android/view/ViewTreeObserver.smali
--- a/smali_classes4/android/view/ViewTreeObserver.smali
+++ b/smali_classes4/android/view/ViewTreeObserver.smali
@@ -19,7 +19,8 @@
         Landroid/view/ViewTreeObserver$OnScrollChangedListener;,
         Landroid/view/ViewTreeObserver$OnComputeInternalInsetsListener;,
         Landroid/view/ViewTreeObserver$InternalInsetsInfo;,
-        Landroid/view/ViewTreeObserver$OnEnterAnimationCompleteListener;
+        Landroid/view/ViewTreeObserver$OnEnterAnimationCompleteListener;,
+        Landroid/view/ViewTreeObserver$WindowInsetsAnimationListener;
     }
 .end annotation
 
@@ -29,6 +30,16 @@
 
 
 # instance fields
+.field private blacklist mWindowInsetsAnimationListeners:Landroid/view/ViewTreeObserver$CopyOnWriteArray;
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "Landroid/view/ViewTreeObserver$CopyOnWriteArray<",
+            "Landroid/view/ViewTreeObserver$WindowInsetsAnimationListener;",
+            ">;"
+        }
+    .end annotation
+.end field
+
 .field private greylist-max-o mAlive:Z
 
 .field private blacklist mGestureExclusionListeners:Landroid/view/ViewTreeObserver$CopyOnWriteArray;
@@ -1574,6 +1585,22 @@
 
     :cond_19
     :goto_c
+    iget-object v0, p1, Landroid/view/ViewTreeObserver;->mWindowInsetsAnimationListeners:Landroid/view/ViewTreeObserver$CopyOnWriteArray;
+
+    if-eqz v0, :unica_insets_merged
+
+    iget-object v1, p0, Landroid/view/ViewTreeObserver;->mWindowInsetsAnimationListeners:Landroid/view/ViewTreeObserver$CopyOnWriteArray;
+
+    if-eqz v1, :unica_transfer_insets_listeners
+
+    invoke-virtual {v1, v0}, Landroid/view/ViewTreeObserver$CopyOnWriteArray;->addAll(Landroid/view/ViewTreeObserver$CopyOnWriteArray;)V
+
+    goto :unica_insets_merged
+
+    :unica_transfer_insets_listeners
+    iput-object v0, p0, Landroid/view/ViewTreeObserver;->mWindowInsetsAnimationListeners:Landroid/view/ViewTreeObserver$CopyOnWriteArray;
+
+    :unica_insets_merged
     invoke-direct {p1}, Landroid/view/ViewTreeObserver;->kill()V
 
     return-void
@@ -1916,3 +1943,171 @@
 
     return p0
 .end method
+
+.method public whitelist addWindowInsetsAnimationListener(Landroid/view/ViewTreeObserver$WindowInsetsAnimationListener;)V
+    .locals 1
+
+    invoke-direct {p0}, Landroid/view/ViewTreeObserver;->checkIsAlive()V
+
+    iget-object v0, p0, Landroid/view/ViewTreeObserver;->mWindowInsetsAnimationListeners:Landroid/view/ViewTreeObserver$CopyOnWriteArray;
+
+    if-nez v0, :cond_0
+
+    new-instance v0, Landroid/view/ViewTreeObserver$CopyOnWriteArray;
+
+    invoke-direct {v0}, Landroid/view/ViewTreeObserver$CopyOnWriteArray;-><init>()V
+
+    iput-object v0, p0, Landroid/view/ViewTreeObserver;->mWindowInsetsAnimationListeners:Landroid/view/ViewTreeObserver$CopyOnWriteArray;
+
+    :cond_0
+    invoke-virtual {v0, p1}, Landroid/view/ViewTreeObserver$CopyOnWriteArray;->add(Ljava/lang/Object;)V
+
+    return-void
+.end method
+
+.method blacklist dispatchOnWindowInsetsAnimationEnd(Landroid/view/WindowInsetsAnimation;)V
+    .locals 4
+
+    iget-object p0, p0, Landroid/view/ViewTreeObserver;->mWindowInsetsAnimationListeners:Landroid/view/ViewTreeObserver$CopyOnWriteArray;
+
+    if-eqz p0, :cond_1
+
+    invoke-virtual {p0}, Landroid/view/ViewTreeObserver$CopyOnWriteArray;->size()I
+
+    move-result v0
+
+    if-lez v0, :cond_1
+
+    const-string/jumbo v0, "VTO#dispatchOnWindowInsetsAnimationEnd"
+
+    invoke-static {v0}, Landroid/os/Trace;->beginSection(Ljava/lang/String;)V
+
+    invoke-virtual {p0}, Landroid/view/ViewTreeObserver$CopyOnWriteArray;->start()Landroid/view/ViewTreeObserver$CopyOnWriteArray$Access;
+
+    move-result-object v0
+
+    :try_start_0
+    invoke-virtual {v0}, Landroid/view/ViewTreeObserver$CopyOnWriteArray$Access;->size()I
+
+    move-result v1
+
+    const/4 v2, 0x0
+
+    :goto_0
+    if-ge v2, v1, :cond_0
+
+    invoke-virtual {v0, v2}, Landroid/view/ViewTreeObserver$CopyOnWriteArray$Access;->get(I)Ljava/lang/Object;
+
+    move-result-object v3
+
+    check-cast v3, Landroid/view/ViewTreeObserver$WindowInsetsAnimationListener;
+
+    invoke-interface {v3, p1}, Landroid/view/ViewTreeObserver$WindowInsetsAnimationListener;->onEnd(Landroid/view/WindowInsetsAnimation;)V
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    add-int/lit8 v2, v2, 0x1
+
+    goto :goto_0
+
+    :cond_0
+    invoke-virtual {p0}, Landroid/view/ViewTreeObserver$CopyOnWriteArray;->end()V
+
+    invoke-static {}, Landroid/os/Trace;->endSection()V
+
+    return-void
+
+    :catchall_0
+    move-exception p1
+
+    invoke-virtual {p0}, Landroid/view/ViewTreeObserver$CopyOnWriteArray;->end()V
+
+    invoke-static {}, Landroid/os/Trace;->endSection()V
+
+    throw p1
+
+    :cond_1
+    return-void
+.end method
+
+.method blacklist dispatchOnWindowInsetsAnimationPrepare(Landroid/view/WindowInsetsAnimation;)V
+    .locals 4
+
+    iget-object p0, p0, Landroid/view/ViewTreeObserver;->mWindowInsetsAnimationListeners:Landroid/view/ViewTreeObserver$CopyOnWriteArray;
+
+    if-eqz p0, :cond_1
+
+    invoke-virtual {p0}, Landroid/view/ViewTreeObserver$CopyOnWriteArray;->size()I
+
+    move-result v0
+
+    if-lez v0, :cond_1
+
+    const-string/jumbo v0, "VTO#dispatchOnWindowInsetsAnimationPrepare"
+
+    invoke-static {v0}, Landroid/os/Trace;->beginSection(Ljava/lang/String;)V
+
+    invoke-virtual {p0}, Landroid/view/ViewTreeObserver$CopyOnWriteArray;->start()Landroid/view/ViewTreeObserver$CopyOnWriteArray$Access;
+
+    move-result-object v0
+
+    :try_start_0
+    invoke-virtual {v0}, Landroid/view/ViewTreeObserver$CopyOnWriteArray$Access;->size()I
+
+    move-result v1
+
+    const/4 v2, 0x0
+
+    :goto_0
+    if-ge v2, v1, :cond_0
+
+    invoke-virtual {v0, v2}, Landroid/view/ViewTreeObserver$CopyOnWriteArray$Access;->get(I)Ljava/lang/Object;
+
+    move-result-object v3
+
+    check-cast v3, Landroid/view/ViewTreeObserver$WindowInsetsAnimationListener;
+
+    invoke-interface {v3, p1}, Landroid/view/ViewTreeObserver$WindowInsetsAnimationListener;->onPrepare(Landroid/view/WindowInsetsAnimation;)V
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    add-int/lit8 v2, v2, 0x1
+
+    goto :goto_0
+
+    :cond_0
+    invoke-virtual {p0}, Landroid/view/ViewTreeObserver$CopyOnWriteArray;->end()V
+
+    invoke-static {}, Landroid/os/Trace;->endSection()V
+
+    return-void
+
+    :catchall_0
+    move-exception p1
+
+    invoke-virtual {p0}, Landroid/view/ViewTreeObserver$CopyOnWriteArray;->end()V
+
+    invoke-static {}, Landroid/os/Trace;->endSection()V
+
+    throw p1
+
+    :cond_1
+    return-void
+.end method
+
+.method public whitelist removeWindowInsetsAnimationListener(Landroid/view/ViewTreeObserver$WindowInsetsAnimationListener;)V
+    .locals 0
+
+    invoke-direct {p0}, Landroid/view/ViewTreeObserver;->checkIsAlive()V
+
+    iget-object p0, p0, Landroid/view/ViewTreeObserver;->mWindowInsetsAnimationListeners:Landroid/view/ViewTreeObserver$CopyOnWriteArray;
+
+    if-nez p0, :cond_0
+
+    return-void
+
+    :cond_0
+    invoke-virtual {p0, p1}, Landroid/view/ViewTreeObserver$CopyOnWriteArray;->remove(Ljava/lang/Object;)V
+
+    return-void
+.end method
